### ISSUE TITLE
[Bug] spatial_softargmax2d tests broken

### ISSUE THREAD

**Issue by edgarriba** [kornia/kornia#535]
In the current version the tests for spatial_softargmax2d tests are broken.

After checking the code I believe that there's a mismatch in the normalization parameter.
https://github.com/kornia/kornia/blob/master/test/geometry/test_spatial_softargmax.py#L77

**Comment by anibali** [comment]
The tests themselves are indeed quite broken, but the code being tested is correct as far as I can see.

The main problem is that the author of these tests confused the similarly named `spatial_softargmax_2d` and `spatial_soft_argmax2d` functions (see https://github.com/kornia/kornia/issues/353#issuecomment-587165931). To recap the difference, `spatial_soft_argmax2d` applies spatial softmax before calling `spatial_softargmax_2d`. Is the naming terrible? Yes, I made a mistake there for sure.

My suggestion would be to rename one of these functions for clarity, or to simply merge the functions and introduce a boolean argument for whether softmax should be applied.

Anyway, https://github.com/kornia/kornia/pull/544 fixes the tests.

**Comment by edgarriba** [comment]
indeed, probably it's worth renaming. In order to make it consistent with [convolutional](https://kornia.readthedocs.io/en/latest/geometry.html?highlight=conv_softargmax#kornia.geometry.conv_soft_argmax2d) versions, what about `spatial_soft_argmax2d` and `spatial_argmax2d` ?

Besides, what about grouping under a same submodule `kornia.geometry.coord` ? (or other related  name) /cc @ducha-aiki

**Comment by anibali** [comment]
The name `spatial_argmax2d` is not good since argmax is the non-differentiable operation that takes the location of the highest-value pixel (ie the "arguments" x, y that give the maximum value). That's how I would expect a function called `spatial_argmax2d` to behave. The operation that we are doing is more like an expectation or weighted mean. How about `spatial_expectation2d`?

**Comment by edgarriba** [comment]
definitely this is confusing, I would go for one single functionality. Why did you split both ? just curious.
For consistency I would have:
- `spatial_softmax2d`
--> does only softmax over 2d, actually this could be a generic thing by specifying a tuple of consecutive axis, right ?
- `spatial_soft_argmax2d`
- `conv_soft_argmax2d`

**Comment by anibali** [comment]
Let's draw a comparison to a similar situation in PyTorch itself. In PyTorch they have both [`nll_loss`](https://pytorch.org/docs/stable/nn.functional.html#nll-loss) and [`cross_entropy`](https://pytorch.org/docs/stable/nn.functional.html#cross-entropy). The first is a meaningful individual operation which expects a log probability input, the second is a convenience which combines [`log_softmax`](https://pytorch.org/docs/stable/nn.functional.html#log-softmax) and `nll_loss` since they are often used together. We have a very similar situation here.

Our "expectation" operation is analogous to the `nll_loss` function. It is commonly used immediately after a softmax, but does not have to be. It simply requires a probability map (pixels are non-negative and sum to 1), which could be obtained through means other than softmax.

So to have a good API for everything (let's leave the convolutional function out of the discussion for now), we should have 3 operations:

* `spatial_soft_argmax2d` (analogy: `cross_entropy`, simply combines the other two)
* `spatial_softmax2d` (analogy: `log_softmax`)
* `spatial_expectation2d`? (analogy: `nll_loss`)

**Comment by edgarriba** [comment]
Perfect, I buy your history and go ahead with the renaming. What about also the module thing?
